Reminder-App/Reminder-App/ReminderListViewController.swift

```diff
@@ -9,32 +9,217 @@ import UIKit
         }
     }
     
     // MARK: Lifecycle Methods
     override func viewDidLoad() {
-        
         super.viewDidLoad()
         
-        reminderListDataSource = ReminderListDataSource()
+        reminderListDataSource = ReminderListDataSource(
+            reminderCompletedAction: {
+                reminderIndex in
+                DispatchQueue.main.async {
+                    self.tableView.reloadRows(at: [IndexPath(row: reminderIndex, section: 0)], with: .automatic)
+                    self.refreshProgressView()
+                }
+            }, reminderDeletedAction: {
+                DispatchQueue.main.async {
+                    self.refreshProgressView()
+                }
+            }, remindersChangedAction: {
+                DispatchQueue.main.async {
+                    self.tableView.reloadData()
+                    self.refreshProgressView()
+                }
+            })
+        
         tableView.dataSource = reminderListDataSource
     }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        
+        let radius = view.bounds.size.width * 0.5 * 0.7
+        progressContainerView.layer.cornerRadius = radius
+        progressContainerView.layer.masksToBounds = true
+        self.refreshProgressView()
+        refreshBackground()
+    }
+    
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        
+        if let navigationController = navigationController,
+           navigationController.isToolbarHidden {
+            navigationController.setToolbarHidden(false, animated: animated)
+        }
+    }
+    
+    // MARK: Private Methods
+    private func addReminder() {
+        let storyboard = UIStoryboard(name: Self.mainStoryboardName, bundle: nil)
+        let detailViewController: ReminderDetailViewController = storyboard.instantiateViewController(identifier: Self.detailViewControllerIdentifier)
+        let reminder = Reminder(id: UUID().uuidString, title: "New Reminder", dueDate: Date())
+        
+        detailViewController.configure(with: reminder,
+                                       isNew: true,
+                                       addAction: {
+                                        reminder in
+                                        self.reminderListDataSource?.add(reminder, completion: {
+                                            (index) in
+                                            if let index = index {
+                                                DispatchQueue.main.async {
+                                                    self.tableView.insertRows(at: [IndexPath(row: index, section: 0)], with: .automatic)
+                                                    self.refreshProgressView()
+                                                }
+                                            }
+                                        })
+                                       })
```

Comment:
There is a lot of nested code here. to avoid nested closures and completion handlers. you can divide the code into different methods..e.g
you can create a method `addReminder(_ reminder: Reminder)` and call that method in `addAction` closure
```
detailViewController.configure(with: reminder,
                                       isNew: true,
                                       addAction: {
                                        reminder in
                                        self.addReminder(reminder)
                                        })
```

Reply:
Also if addActions is trailing closure you can omit its name